Intermittent problem when opening database

### Discussed in https://github.com/eclipse-store/store/discussions/263

<div type='discussions-op-text'>

<sup>Originally posted by **HansITChange** August  1, 2024</sup>
I'm running version 1.4.0, however this problem was also present in previous versions. Now and then, I get following error:
```console
Error occurred in storage channel#0
org.eclipse.store.storage.exceptions.StorageExceptionIoWriting: null
	at org.eclipse.store.storage.types.StorageFileWriter.truncateFile(StorageFileWriter.java:168)
	at org.eclipse.store.storage.types.StorageFileWriter.truncate(StorageFileWriter.java:144)
	at org.eclipse.store.storage.types.StorageFileManager$Default.handleLastFile(StorageFileManager.java:1304)
	at org.eclipse.store.storage.types.StorageFileManager$Default.initializeForExistingFiles(StorageFileManager.java:1011)
	at org.eclipse.store.storage.types.StorageFileManager$Default.initializeStorage(StorageFileManager.java:886)
	at org.eclipse.store.storage.types.StorageChannel$Default.initializeStorage(StorageChannel.java:782)
	at org.eclipse.store.storage.types.StorageChannelTaskInitialize$Default.succeed(StorageChannelTaskInitialize.java:200)
	at org.eclipse.store.storage.types.StorageChannelTaskInitialize$Default.succeed(StorageChannelTaskInitialize.java:34)
	at org.eclipse.store.storage.types.StorageChannelSynchronizingTask$AbstractCompletingTask.synchronizedComplete(StorageChannelSynchronizingTask.java:78)
	at org.eclipse.store.storage.types.StorageChannelSynchronizingTask$AbstractCompletingTask.complete(StorageChannelSynchronizingTask.java:126)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.processBy(StorageChannelTask.java:260)
	at org.eclipse.store.storage.types.StorageChannel$Default.work(StorageChannel.java:453)
	at org.eclipse.store.storage.types.StorageChannel$Default.run(StorageChannel.java:536)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.eclipse.store.storage.exceptions.StorageExceptionConsistency: Reopened file has inconsistent size: Expected size 50962 != actual size 51066.
	at org.eclipse.store.storage.types.StorageLiveDataFile$Default.internalOpenWriting(StorageLiveDataFile.java:282)
	at org.eclipse.store.storage.types.StorageFile$Abstract.ensureWritable(StorageFile.java:390)
	at org.eclipse.store.storage.types.StorageFile$Abstract.truncate(StorageFile.java:366)
	at org.eclipse.store.storage.types.StorageFileWriter.truncateFile(StorageFileWriter.java:164)
```

Where is this coming from ? 
And how can I recover from this error ? The only way I found was to empty the database, and start all over... Is there a more intelligent way to recover from this error ?</div>

--- a/storage/storage/src/main/java/org/eclipse/store/storage/types/StorageFile.java
+++ b/storage/storage/src/main/java/org/eclipse/store/storage/types/StorageFile.java
@@ -361,7 +361,7 @@ public interface StorageFile
 			}
 		}
 				
-		public final synchronized void truncate(final long newLength)
+		public synchronized void truncate(final long newLength)
 		{
 			this.ensureWritable().truncate(newLength);
 		}

--- a/storage/storage/src/main/java/org/eclipse/store/storage/types/StorageLiveDataFile.java
+++ b/storage/storage/src/main/java/org/eclipse/store/storage/types/StorageLiveDataFile.java
@@ -153,6 +153,15 @@ extends StorageDataFile, StorageLiveChannelFile<StorageLiveDataFile>, StorageCre
 			return this.fileTotalLength;
 		}
 		
+		@Override
+		public synchronized void truncate(long newLength)
+		{
+			//Do not use this.internalOpenWriting() because this method does an size check
+			//if the file is re-opened.
+			//This size check is not appropriate if the file shall be opened with write access for truncation.
+			super.internalOpenWriting();
+			super.truncate(newLength);
+		}
 
 		final TypeInFile typeInFile(final StorageEntityType.Default type)
 		{
@@ -287,7 +296,7 @@ extends StorageDataFile, StorageLiveChannelFile<StorageLiveDataFile>, StorageCre
 			
 			return true;
 		}
-
+		
 		@Override
 		public final double dataFillRatio()
 		{

